chapter6.adoc

```diff
@@ -67,24 +67,85 @@ response to the address translation request on the device translation completion
+The MTT access permission lookup process used by MTTCHK is identical to that
+specified by Smmtt extension in section "MTT access permissions lookup process".
+
+The RISC-V memory model requires memory access from a hart to be single-copy
+atomic. When RV32 is implemented the size of a single-copy atomic memory access
+is up to 32-bits. When RV64 is implemented the size of a single-copy atomic
+memory access is up to 64-bits. The size of a single-copy atomic memory access
+implemented by MTTCHK is UNSPECIFIED but is required to be at least 32-bits if
+all of the harts in the system implement RV32 and is required to be at least
+64-bits if any of the harts in the system implement RV64. Software must follow
+the rules outlined below to update MTT entries.
+
+* It is generally unsafe to update fields of an MTT entry using stores of width
+  less than the minimal single-copy atomic memory access supported by MTTCHK as
+  it is legal for MTTCHK to read the entry at any time, including when only
+  some of the partial stores have taken effect. For an update to be atomic,
+  software must use a single store of width equal to the minimal single-copy
+  atomic memory access supported by MTTCHK.
+
+* MTTCHK is not required to immediately observe the software updates to an MTT
+  entry. Software must use the `MTTINVAL` operation outlined in <<section>> to
+  invalidate any previous copies of that entry that may be in the MTTCHK caches
+  to synchronize the updates to the entry with the operation of MTTCHK.
 
-Each IOMMU offers a memory-mapped register programming interface that the
-associated supervisor domain uses to configure and control the IOMMU. The RISC-V
-hart employs the MTT to enforce access controls on the physical address. This is
-done to ascertain whether the supervisor domain currently executing on that
-RISC-V hart possesses the rights to access the physical address of an IOMMU
-register programming interface.
+[NOTE]
+====
+If an MTT entry is changed, MTTCHK may use the old value of the entry or the
+new value of the entry and the choice is unpredictable until software uses the
+`MTTINVAL` operation to synchronize updates to the entry with the operation of
+the MTTCHK. These are the only behaviors expected.
+====
 
-The I/O MTT checker provides a memory-mapped register programming interface
-associated with the RDSM. The RDSM employs the MTT (or PMP) to prohibit access
-to these registers from any of the supervisor domains.
+The I/O MTT checker provides a memory-mapped register programming interface that
+is expected to be associated with the RDSM. The RDSM may employ the MTT (or PMP)
+to prohibit access to these registers from any of the supervisor domains.
+
+If an MTT check disallows a transaction then the transaction is aborted.
+
+If the aborted transaction is an IOMMU-initiated memory access then the IO bridge
+signals such access faults to the IOMMU itself. The details of such signaling is
+implementation defined.
```

Comment:
On a related note: is it expected that IOMMU-initiated transactions (i.e. page table walks) go through the MTTCHK? If so, we should specify this.

Reply:
It's stated here:
> The MTTCHK uses the SDID to determine the MTT associated with the supervisor domain and checks if the physical address may be accessed by the device or IOMMU associated with that supervisor domain.